core/sail/extensible-store/src/main/java/org/eclipse/rdf4j/sail/extensiblestore/evaluationstatistics/ExtensibleDynamicEvaluationStatistics.java

```diff
@@ -0,0 +1,337 @@
+		Stream.of(subjectPredicateIndex, predicateObjectIndex, subjectPredicateIndex_removed,
+				predicateObjectIndex_removed).forEach(index -> {
+					for (int i = 0; i < index.length; i++) {
+						for (int j = 0; j < index[i].length; j++) {
+							index[i][j] = HyperLogLogCollector.makeLatestCollector();
+						}
+					}
+				});
+
+	}
+
+	@Override
+	protected CardinalityCalculator createCardinalityCalculator() {
+		return cardinalityCalculator;
+	}
+
+	CardinalityCalculator cardinalityCalculator = new CardinalityCalculator() {
+
+		@Override
+		protected double getCardinality(StatementPattern sp) {
+			double min = size.estimateCardinality() - size_removed.estimateCardinality();
+
+			min = Math.min(min, getSubjectCardinality(sp.getSubjectVar()));
+			min = Math.min(min, getPredicateCardinality(sp.getPredicateVar()));
+			min = Math.min(min, getObjectCardinality(sp.getObjectVar()));
+
+			if (sp.getSubjectVar().getValue() != null && sp.getPredicateVar().getValue() != null) {
+				min = Math.min(min,
+						getHllCardinality(subjectPredicateIndex, subjectPredicateIndex_removed,
+								sp.getSubjectVar().getValue(),
+								sp.getPredicateVar().getValue()));
+			}
+
+			if (sp.getPredicateVar().getValue() != null && sp.getObjectVar().getValue() != null) {
+				min = Math.min(min,
+						getHllCardinality(predicateObjectIndex, predicateObjectIndex_removed,
+								sp.getPredicateVar().getValue(),
+								sp.getObjectVar().getValue()));
+			}
+
+			return min;
+
+		}
+
+		@Override
+		protected double getSubjectCardinality(Var var) {
+			if (var.getValue() == null) {
+				return size.estimateCardinality();
+			} else {
+				return getHllCardinality(subjectIndex, subjectIndex_removed, var.getValue());
+			}
+
+		}
+
+		@Override
+		protected double getPredicateCardinality(Var var) {
+			if (var.getValue() == null) {
+				return size.estimateCardinality();
+			} else {
+				return getHllCardinality(predicateIndex, predicateIndex_removed, var.getValue());
+			}
+		}
+
+		@Override
+		protected double getObjectCardinality(Var var) {
+			if (var.getValue() == null) {
+				return size.estimateCardinality();
+			} else {
+				return getHllCardinality(objectIndex, objectIndex_removed, var.getValue());
+			}
+		}
+
+		@Override
+		protected double getContextCardinality(Var var) {
+			if (var.getValue() == null) {
+				return defaultContext.estimateCardinality() - defaultContext_removed.estimateCardinality();
+			} else {
+				return getHllCardinality(contextIndex, contextIndex_removed, var.getValue());
+			}
+		}
+	};
+
+	private double getHllCardinality(HyperLogLogCollector[][] index, HyperLogLogCollector[][] index_removed,
+			Value value1, Value value2) {
+		return index[Math.abs(value1.hashCode() % index.length)][Math.abs(value2.hashCode() % index.length)]
+				.estimateCardinality()
+				- index_removed[Math.abs(value1.hashCode() % index_removed.length)][Math
+						.abs(value2.hashCode() % index_removed.length)]
+								.estimateCardinality();
```

Comment:
I'll have to believe you if you say that it does what it should do. Seriously though: I wonder if you could maybe split this out into more than one line of code to make it easier to read.

Reply:
Will do :D - it got a lot worse when I added in the index_removed support.